Reshape adding support for RM tensors with different widths, tile tensors with padding, tensors with 1 element on inner dim after reshape

ntarafdar opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/12715
https://github.com/tenstorrent/tt-metal/issues/12866
https://github.com/tenstorrent/tt-metal/issues/13159
https://github.com/tenstorrent/tt-metal/issues/12644

### Problem description
Reshape has several issues when it comes to functional generality.
This reshape operation was previously in core.cpp and implements reshape by modifying the view of the tensor. 
This has also been moved to the data_movement folder.

The focus of this PR is to address the missing 3 things below. We will investigate full generality in a future PR

Missing cases, tensors more than rank 4.

1. For RM Layout Tensors, the inner dimension needs to be same (Asserts otherwise)
2. For Tile Tensors the tiled dimensions cannot be padded (asserts otherwise)
3. If we redirect tensors in case 1 to our Reshape_on_device operation, this will work , this performs the transpose required if the inner dimensions change. But that kernel has a limitation where tensors need to be aligned to 32bit width. (this is a limitation in most operations tbh) 


### What's changed
1. Redirect RM tensors with different inner dimensions to the reshape_on_device operation
2. For Tile tensors, untilize and remove padding then use the RM tensor path above
3. If an RM tensors inner dimension is not 32 bit aligned, fall back to host. 

### Checklist
- [x] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/11203161729)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

ntarafdar committed:
#12715: fix row-major reshape with different page-widths

tarafdarTT commented:
Passing Post-Commit
https://github.com/tenstorrent/tt-metal/actions/runs/11017346368

ayerofieiev-tt commented:
What is "diff page"?

tarafdarTT commented:
> What is "diff page"?

for row major tensors the width of the page in memory is the width of the row. It asserts out currently if the page width is different between the input and output.

ayerofieiev-tt commented on tests/ttnn/unit_tests/test_reshape.py:
```diff
@@ -285,3 +285,16 @@ def test_reshape_tile_layout_only_change_shape(device):
     assert_with_pcc(torch_result, output, 0.9999)
+
+
+def test_reshape_bert(device):
```
can we call it smth different? not just bert, but whats the semantics. 
test_reshape_rowmajor_change_row_size

ntarafdar committed:
#0: move reshape from core to reshape_view

#0: update reshape

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,123 @@
+    return detail::row_major_reshape(tensor, shape);
+
+
+}
```
1. In this function there are 3 calls to `row_major_reshape` alone. I count 6 calls to `tensor.reshape(shape.value)` across 3 functions. Besides that there is one call to `..., data_transfer_to_device` and one `to_layout`

Could the code be fixed up to something look like the following (I don't have preferences for if vs else if and whether helper functions are used to evaluate the booleans)?
```
    if (tensor_shape == shape) {
        return {tensor};
    }
    if (do_to_layout) {
        ...
        return to_layout...;
   }
    if (do_device) {
        ...
        return data_transfer_to_device...;
   }
   return tensor.reshape(shape.value)
```

2. The original code had a path were reshape was not handled. With this change, will all possible shapes be handled? Or can there be a potential downstream assert/fatal that would be triggered?

ayerofieiev-tt replied:
Agree

ntarafdar replied:
Thanks @bbradelTT I cleaned it up, the only case that I know we can't handle at the moment is more than 4D tensors, for which I added a FATAL

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:
```diff
@@ -0,0 +1,124 @@
+        module,
+        ttnn::reshape,
+
+        R"doc(reshape(input_tensor: ttnn.Tensor, ttnn.Shape new_shape) -> ttnn.Tensor
```
After https://github.com/tenstorrent/tt-metal/pull/12075 you're not supposed to have the header.

yan-zaretskiy replied:
Also, do not use `:attr:` directives, just take a look at other ops to see what the new docs format is.

ntarafdar committed:
#0: update kernels

Update ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

Co-authored-by: Artem Yerofieiev <169092593+ayerofieiev-tt@users.noreply.github.com>

Update ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

Co-authored-by: Artem Yerofieiev <169092593+ayerofieiev-tt@users.noreply.github.com>

ayerofieiev-tt commented on tests/ttnn/unit_tests/test_reshape.py:
```diff
@@ -285,3 +285,45 @@ def test_reshape_tile_layout_only_change_shape(device):
+    torch_input_tensor = torch.randn(input_shape, dtype=torch.bfloat16)
+    torch_result = torch_input_tensor.reshape(output_shape)
+
+    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
```
this test will fail if attempted in Row Major, right?

ntarafdar replied:
it shouldn't as the Row major path is there.
but ya that's my bad on the test, I'll parameterize the input layout to cover both.

ntarafdar replied:
Actually, I separated the two tests, actually one for tensors ending in shape 1 , and one for without.
The ones not ending in 1 I can do both TILE_LAYOUT and ROW_MAJOR_LAYOUT/

You are right this will fail but that's not due to reshape. it will fail in the creation of the torch tensor independent of reshape, hence this has to be in TILE_LAYOUT

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:
```diff
@@ -0,0 +1,124 @@
+        ttnn::pybind_overload_t{
+            [](const data_movement_operation_t& self,
+               const ttnn::Tensor& input_tensor,
+               const std::array<uint32_t, 1>& new_shape
```
instead of 5 overloads, would it be better to have 1 that accepts vector?

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,134 @@
+        return tensor;
+    }
+
+    if (layout == ttnn::Layout::ROW_MAJOR) {
+        return detail::row_major_reshape(tensor, shape);
+    }
+    else {
+        const auto new_shape_with_tile_padding = shape.with_tile_padding();
+        const auto new_height = new_shape_with_tile_padding[-2];
+        const auto new_width = new_shape_with_tile_padding[-1];
+
+        const auto is_tile_multiple = (new_height % ttnn::TILE_SIZE == 0 && new_width % ttnn::TILE_SIZE == 0);
+        if (not is_tile_multiple) {
+            return detail::row_major_reshape(tensor, shape);
+        }
+
+        if (ttnn::has_storage_type_of(tensor, ttnn::StorageType::DEVICE)) {
+            if (tensor_shape.with_tile_padding()[-1] == new_width) {
+                return tensor.reshape(shape.value);
+            }
+        } else {
+            return tensor.reshape(shape.value);
+        }
+    }
+    return detail::row_major_reshape(tensor, shape);
```
```suggestion
    if (layout == ttnn::Layout::ROW_MAJOR) {
        return detail::row_major_reshape(tensor, shape);
    }
    
    const auto new_shape_with_tile_padding = shape.with_tile_padding();
    const auto new_height = new_shape_with_tile_padding[-2];
    const auto new_width = new_shape_with_tile_padding[-1];

    const auto is_tile_multiple = (new_height % ttnn::TILE_SIZE == 0 && new_width % ttnn::TILE_SIZE == 0);
    if (not is_tile_multiple) {
        return detail::row_major_reshape(tensor, shape);
    }

    if (!ttnn::has_storage_type_of(tensor, ttnn::StorageType::DEVICE)) {
        return tensor.reshape(shape.value);
    }
    
    if (tensor_shape.with_tile_padding()[-1] == new_width) {
        return tensor.reshape(shape.value);
    }

    return detail::row_major_reshape(tensor, shape);
```

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,134 @@
+            }
+        }
+    } else if (tensor_shape.rank() >= 2 and shape.rank() >= 2) {
+        // Handle the case when the tensor is not contiguous but the last two dimensions are the same and so reshape
```
is there a test for this case?

ntarafdar replied:
There is not, I don't think this test is condition is necessary and am unsure how to recreate this (A RM tensor with padding different than legacy shape, I think will make more sense when we pad up to 32 bit alignment with the new tensor) , but I am hesitant to remove it as this is part of the legacy reshape that was in core.cpp before I inherited it.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,134 @@
+                return  host_reshape(rm_tensor, shape);
+            }
+            else {
+                auto tensor_4d = unsqueeze_to_4D(rm_tensor);
```
what does this do for 5d inputs?

ntarafdar replied:
It does not work since slice doesn't work for 5 inputs :( , I updated the description of the PR to reflect that.

ayerofieiev-tt replied:
Thank you!
Lets add a test which we expect won't work today with a short note on why

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,168 @@
+    } else {
+        return detail::row_major_reshape(tensor, shape);
+    }
+
+
+
+
+}
+
+
```
Please avoid adding these such random indentations

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,168 @@
+    const std::vector<int32_t> & shape_vector
+    ) {
+
+    auto shape = detail::get_shape_from_vector(tensor, shape_vector);
```
shape_vector is a logical shape. it is supposed to be like that and it makes no sense creating full shape with padding from tensor, it is error prone

ntarafdar replied:
this is to handle negative values for reshape. its the same code that is in core.hpp 
without this we won't be able to handle negative values.

ayerofieiev-tt replied:
Oh, learned something new. Thank you!
Can you add a test with a negative value?

ntarafdar replied:
there are two tests already for negative values, I copied what's in core.hpp (and will delete the old code in core.hpp) to avoid regression on the negative value tests. 
https://github.com/tenstorrent/tt-metal/blob/tarafdarTT/reshape_fix/tests/ttnn/unit_tests/test_reshape.py#L241 
https://github.com/tenstorrent/tt-metal/blob/tarafdarTT/reshape_fix/tests/ttnn/unit_tests/test_reshape.py#L176

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:
```diff
@@ -0,0 +1,72 @@
+        R"doc(
+
+        Args:
+            * :attr:`input_tensor`: Input Tensor.
```
please get docs guide from @yan-zaretskiy , I think :attr: is not needed anymore

ntarafdar replied:
ah true, sorry forgot to push that commit

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,160 @@
+
+ttnn::Tensor ReshapeViewOperation::invoke(
+    const ttnn::Tensor& tensor,
+    const ttnn::Shape& shape
```
Why don't you call `detail::get_shape_from_vector(tensor, shape_vector)` here?

ntarafdar replied:
I think ttnn::Shape cannot contain negatives as the underlying LegacyShape that it uses for its value uses std::array<uint32_t, MAX_NUM_DIMENSIONS>

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,160 @@
+    std::vector<std::uint32_t> new_shape(shape.size());
+    std::copy(shape.begin(), shape.end(), new_shape.begin());
+    if (new_volume < 0) {
+        const auto volume = tensor.get_shape().with_tile_padding().volume();
```
Why is it with_tile_padding that matters for volume and not the logical volume?
My expectation is that its logical volume that matters

ntarafdar replied:
That's true, it should be, I'll have to add more testing for negative values (outside scope of the PR). 
I renamed the PR to explicitly list what its doing

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,160 @@
+    for (auto index = 0; index < shape.size(); ++index) {
+        if (shape[index] == -1) {
+            if (index_of_negative_1 != -1) {
+                TT_THROW("Shape cannot have more than 1 elements that is set to -1!");
```
please add shape to the error message

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -0,0 +1,160 @@
+
+}
+
+ttnn::Shape get_shape_from_vector(const ttnn::Tensor& tensor, const std::vector<int32_t> & shape) {
```
please give this method a different name to reflect what it really does

ayerofieiev-tt commented on tests/ttnn/unit_tests/test_reshape.py:
```diff
@@ -285,3 +281,46 @@ def test_reshape_tile_layout_only_change_shape(device):
+    [
+        ((1, 256, 1), (1, 256)),
+        ((1, 1024, 1), (1, 4, 256)),
+        ((1, 128, 1), (1, 128)),
```
lets add 1 test with -1?

ayerofieiev-tt commented on tests/ttnn/unit_tests/test_reshape.py:
```diff
@@ -285,3 +281,46 @@ def test_reshape_tile_layout_only_change_shape(device):
+
+
+# Reshape in Tile layout with shapes that are not divisible by 32
+@pytest.mark.parametrize(
```
Thinking of more test cases like these
((1, 1, 1, 16, 2), (1, 1, 1, 32)),
((1, 1, 1, 4, 4), (1, -1, 4)),
((1, 1, 1, 8, 8, 1024), (1, 64, 1024))


from here
https://github.com/tenstorrent/pytorch2.0_ttnn/blob/kw/metrics-support-more-statuses/docs/operations/aten.view.default.md

https://github.com/tenstorrent/pytorch2.0_ttnn/blob/kw/metrics-support-more-statuses/docs/operations/aten._unsafe_view.default.md

ntarafdar replied:
I think at the moment, we cannot support any of these because the inputs are more than rank 4. 

I can add this support in a subsequent PR.

eyonland reviewed (approved):
Would love to see C++ tests for reshape.   I think this would make things easier when the tensor layout work is underway.
